### PR TITLE
Index imported modules

### PR DESCRIPTION
**Pull request by ivan444**
- Emit Kythe schema for imported modules (using ref/imports)
- Fix testRecursiveRef which was broken in recent commits (please verify that the test is reflecting the desired behavior)

Issue: https://github.com/google/haskell-indexer/issues/6

### CODE REVIEW

File: haskell-indexer-backend-ghc/tests/Language/Haskell/Indexer/Backend/Ghc/Test/BasicTestBase.hs [hunk 1]
```diff
@@ -50,7 +50,11 @@ testRecursiveRef = assertXRefsFrom "basic/RecursiveRef.hs" $ do
         _ -> checking $ assertFailure "Usage count differs"
     -- Recursive fun with type signature should target using the
     -- polymorphic binding.
-    declAt (11,1) >>= singleUsage >>= includesPos (11,16)
+    declAt (11,1) >>= usages >>= \case
+        [u1, u2] -> do
+            includesPos (10,1) u1
```
Comment (robinp) [inline haskell-indexer-backend-ghc/tests/Language/Haskell/Indexer/Backend/Ghc/Test/BasicTestBase.hs]: Thanks! This is due to the newly emitted `ref/doc` reference from the type signature - can add a Note here, later we might sort if we actually want these to be counted in 'usage'.

File: haskell-indexer-translate/src/Language/Haskell/Indexer/Translate.hs [hunk 2]
```diff
@@ -238,6 +240,11 @@ data Relation = Relation
 data RelationKind = ImplementsMethod | InstantiatesClass
     deriving (Eq, Ord, Show)
 
+-- | Module import
+data Import = Import
```
Comment (robinp) [inline haskell-indexer-translate/src/Language/Haskell/Indexer/Translate.hs]: Can we just use the `ModuleTick` instead? It would contain the pkg+module, as well as the Span (of the import here).

File: kythe-verification/testdata/basic/ImportsRef.hs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+-- Verify the references to the imported modules
```
Comment (robinp) [inline kythe-verification/testdata/basic/ImportsRef.hs]: The current way won't actually crosslink I fear, since when we emit a module entry, the package is part of its signature (see `ModuleTick`). See comment above. After fixing, please add a VName verification on `DS` to check that the package is also included. See `Module.hs` in the verifier testdata on how to do VName asserts.

**Comment by robinp** [comment]
Thanks, will have a look! FYI Travis CI is unhappy about GHC 8 targets.

**Comment by ivan444** [comment]
It's not done yet :-). Just merged the main repo. I'll try to fix the Ghc 8
issue tonight.

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs [hunk 4]
```diff
@@ -440,6 +441,33 @@ relationsFromRenamed ctx declAlts (hsGroup, _, _, _) =
...
+-- | Exports module imports.
+importsFromRenamed :: ExtractCtx -> RenamedSource -> [Import]
+importsFromRenamed ctx (_, lImportDecls, _, _) = map mkImport lImportDecls
+  where
+    mkTick :: ImportDecl Name -> Tick
+    mkTick implDecl = Tick
+        { tickSourcePath = ecSourcePath ctx
+        , tickPkgModule = PkgModule
+            { getPackage = ""
+              -- ^ Unfortunately, GHC AST doesn't fully qualify the import and
```
Comment (ivan444) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs]: Added the package.

File: kythe-verification/testdata/basic/ImportsRef.hs (hunk 3 above)
Comment (ivan444) [inline kythe-verification/testdata/basic/ImportsRef.hs]: Done.

But I'm failing to add vname verification on base:Prelude import. It's not crucial as there's no visible token to bind it, but it would be nice to assert it.

File: haskell-indexer-translate/src/Language/Haskell/Indexer/Translate.hs (hunk 2 above)
Comment (ivan444) [inline haskell-indexer-translate/src/Language/Haskell/Indexer/Translate.hs]: Done.

**Comment by ivan444** [comment]
Please take another look :).

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs [hunk 5]
```diff
@@ -110,33 +110,34 @@ moduleUnitId = modulePackageKey
...
         -- Analysed modules always arrive as file references in practice.
-        modFile = T.pack $!
+    let modFile = T.pack $!
                       fromMaybe "?" (ml_hs_file . ms_location $ modSummary)
-        strippedModFile = SourcePath (aoFilePathTransform opts modFile)
-        ctx = ExtractCtx (ms_mod modSummary) strippedModFile ghcEnv opts
-        renSource = tm_renamed_source tm  :: Maybe RenamedSource
-        tcSource = tm_typechecked_source tm
-        declsAlts =
+    let strippedModFile = SourcePath (aoFilePathTransform opts modFile)
```
Comment (robinp) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs]: One `let` should be enough at start of let-block (even in do-notation)

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs [hunk 6]
```diff
@@ -543,6 +544,24 @@ relationsFromRenamed ctx declAlts (hsGroup, _, _, _) =
...
+    mkImport (L _ implDecl) = do
+      pkgModule <- extractPkgModule . unLoc . ideclName $ implDecl
+      let pkgSpan = give ctx (srcSpanToSpan . getLoc $ ideclName implDecl)
+      return $ ModuleTick pkgModule pkgSpan
+
+    extractPkgModule :: GhcMonad m => ModuleName -> m PkgModule
+    extractPkgModule name = do
+      m <- findModule name Nothing
+      let pkg = T.pack . showPackageName . moduleUnitId $ m
+      let mName = T.pack . moduleNameString . moduleName $ m
```
Comment (robinp) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs]: also here

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs [hunk 7]
```diff
@@ -543,6 +544,24 @@ relationsFromRenamed ctx declAlts (hsGroup, _, _, _) =
...
+  where
+    mkImport :: GhcMonad m => LImportDecl Name -> m ModuleTick
+    mkImport (L _ implDecl) = do
+      pkgModule <- extractPkgModule . unLoc . ideclName $ implDecl
+      let pkgSpan = give ctx (srcSpanToSpan . getLoc $ ideclName implDecl)
+      return $ ModuleTick pkgModule pkgSpan
+
+    extractPkgModule :: GhcMonad m => ModuleName -> m PkgModule
+    extractPkgModule name = do
+      m <- findModule name Nothing
```
Comment (robinp) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs]: Please reuse [`extractModuleName`](https://github.com/google/haskell-indexer/blob/a9eb13947774419e29bc1128d91d13a1558665a6/haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs#L1191) somehow, to use the shared parts and have the package demangling applied.

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/GhcApiSupport.hs [hunk 8]
```diff
@@ -49,16 +52,16 @@ printErr = liftIO . hPutStrLn stderr
 -- can't process multiple compilations concurrently (see
 -- https://mail.haskell.org/pipermail/ghc-devs/2014-January/003874.html).
 withTypechecked
-    :: MVar () -> GhcArgs -> (GhcEnv -> TypecheckedModule -> IO ()) -> IO ()
-withTypechecked globalLock GhcArgs{..} action
+  :: MVar () -> GhcArgs -> AnalysisOptions -> (XRef -> IO ()) -> IO ()
```
Comment (robinp) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/GhcApiSupport.hs]: Please keep 4-indent

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/GhcApiSupport.hs [hunk 9]
```diff
@@ -49,16 +52,16 @@ printErr = liftIO . hPutStrLn stderr
...
+withTypechecked globalLock GhcArgs{..} analysisOpts action
```
Comment (robinp) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/GhcApiSupport.hs]: consider renaming `action` to `xrefSink` or similar.

File: haskell-indexer-backend-ghc/tests/Language/Haskell/Indexer/Backend/Ghc/Test/BasicTestBase.hs [hunk 10]
```diff
@@ -286,6 +286,12 @@ testUtf8 = assertXRefsFrom "basic/Utf8.hs" $
         xs -> checking $ assertFailure
             ("Different use count for unicodey var: " ++ show (length xs))
 
+-- | Test that the module imports are emitted.
+testImports :: AssertionInEnv
+testImports = assertXRefsFrom "basic/Imports.hs" $ do
+    hasImportAt "Data.Text" (3, 8)
```
Comment (robinp) [inline haskell-indexer-backend-ghc/tests/Language/Haskell/Indexer/Backend/Ghc/Test/BasicTestBase.hs]: nit: maybe flipped arg order is more natural to read?

File: haskell-indexer-backend-ghc/tests/Language/Haskell/Indexer/Backend/Ghc/Test/TranslateAssert.hs [hunk 11]
```diff
@@ -107,6 +109,24 @@ prettyReference ref = "TickReference {"
                    ++ ", pos: " ++ prettySpan (refSourceSpan ref)
                    ++ "}"
 
+-- | Returns the module imports that include the given position.
+-- Fails if there aren't any imports, or there are multiple imports.
+-- Note: this doesn't consider the file, so there might be conflicts
+-- if multiple files are processed. Use only for tests where a single file
+-- is loaded, or explicitly test the file of the returned Import.
+hasImportAt :: String -> (Int,Int) -> ReaderT XRef IO ()
```
Comment (robinp) [inline haskell-indexer-backend-ghc/tests/Language/Haskell/Indexer/Backend/Ghc/Test/TranslateAssert.hs]: Please use `Text` instead `String` if possible/reasonable.

File: haskell-indexer-frontend-kythe/src/Language/Haskell/Indexer/Frontend/Kythe.hs [hunk 12]
```diff
@@ -206,6 +207,15 @@ makeRelationFacts (Relation src relKind target) =
            -- So by squinting a bit, we'll hopefully get better functionality
            -- out of Kythe-related tools that are not aware of Haskell.
 
+-- | Makes fact about module imports.
+makeImportFacts :: ModuleTick -> Conversion [Raw.Entry]
+makeImportFacts ModuleTick{..} = do
+    vn <- asks baseVName
+    makeAnchor mtSpan RefImportsE (pkgvn vn) Nothing Nothing
```
Comment (robinp) [inline haskell-indexer-frontend-kythe/src/Language/Haskell/Indexer/Frontend/Kythe.hs]: please use the name portion of `makePackageFacts` (or factor it) to get the package vname.

File: kythe-verification/testdata/basic/ImportsRef.hs [hunk 13]
```diff
@@ -0,0 +1,7 @@
...
+module Imports where
+
+-- - @"Data.Set" ref/imports vname("containers-0.5.7.1:Data.Set", "", "", "", "haskell")
```
Comment (robinp) [inline kythe-verification/testdata/basic/ImportsRef.hs]: Note: Using the vname verification like below with verbatim package name might fail depending on the actual package name mangling. But ok for now, will see if causes any problems.

File: kythe-verification/testdata/basic/ImportsRef.hs (hunk 3 above)
Comment (robinp) [inline kythe-verification/testdata/basic/ImportsRef.hs]: Alternative idea: does Kythe verifier unify the node names across multiple source files when running on multi sources? Then could add a two-file test, and verify that file A exports ModuleA, and file B imports ModuleA (could make a negative test to find out if this works, with three files: B imports A, but verifier asserts it imports C - it should fail). Maybe.

Comment (ivan444) [inline kythe-verification/testdata/basic/ImportsRef.hs]: I like the idea. Resolved the TODO from CrossRef2 test.
Not sure if a negative test is useful. It's more like verifying Kythe framework than haskell-indexer logic.

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs (hunk 5 above)
Comment (ivan444) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs]: Done

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs (hunk 7 above)
Comment (ivan444) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs]: Done.

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs (hunk 6 above)
Comment (ivan444) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs]: Done.

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/GhcApiSupport.hs (hunk 8 above)
Comment (ivan444) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/GhcApiSupport.hs]: Done.

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/GhcApiSupport.hs (hunk 9 above)
Comment (ivan444) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/GhcApiSupport.hs]: Done.

File: haskell-indexer-backend-ghc/tests/Language/Haskell/Indexer/Backend/Ghc/Test/BasicTestBase.hs (hunk 10 above)
Comment (ivan444) [inline haskell-indexer-backend-ghc/tests/Language/Haskell/Indexer/Backend/Ghc/Test/BasicTestBase.hs]: Done.

File: haskell-indexer-backend-ghc/tests/Language/Haskell/Indexer/Backend/Ghc/Test/TranslateAssert.hs (hunk 11 above)
Comment (ivan444) [inline haskell-indexer-backend-ghc/tests/Language/Haskell/Indexer/Backend/Ghc/Test/TranslateAssert.hs]: Done.

File: haskell-indexer-frontend-kythe/src/Language/Haskell/Indexer/Frontend/Kythe.hs (hunk 12 above)
Comment (ivan444) [inline haskell-indexer-frontend-kythe/src/Language/Haskell/Indexer/Frontend/Kythe.hs]: Done.

File: kythe-verification/testdata/basic/ImportsRef.hs (hunk 13 above)
Comment (ivan444) [inline kythe-verification/testdata/basic/ImportsRef.hs]: I'm aware of it. Just couldn't find anything non-versioned except base. We can always switch to a base import if the test turns out to be noisy.

File: haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs [hunk 14]
```diff
@@ -543,6 +544,22 @@ relationsFromRenamed ctx declAlts (hsGroup, _, _, _) =
...
+importsFromRenamed ctx (_, lImportDecls, _, _) = mapM mkImport lImportDecls
+  where
+    mkImport :: GhcMonad m => LImportDecl Name -> m ModuleTick
+    mkImport (L _ implDecl) = do
+      pkgModule <- (extractPkgModule ctx) . unLoc . ideclName $ implDecl
+      let pkgSpan = give ctx (srcSpanToSpan . getLoc $ ideclName implDecl)
+      return $ ModuleTick pkgModule pkgSpan
+
+    extractPkgModule :: GhcMonad m => ExtractCtx -> ModuleName -> m PkgModule
+    extractPkgModule ctx name = do
```
Comment (ivan444) [inline haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/Ghc.hs]: Done.
